lib/iris/_concatenate.py

```diff
@@ -281,14 +288,135 @@ class _CoordExtent(namedtuple("CoordExtent", ["points", "bounds"])):
+        return da.reduction(
+            a,
+            chunk=_chunk,
+            combine=_chunk,
+            aggregate=_aggregate,
+            keepdims=False,
+            meta=np.empty(tuple(), dtype=np.int64),
+            dtype=np.int64,
+        )
+    return _aggregate(a, 0, False)
+
+
+class _ArrayHash(namedtuple("ArrayHash", ["value", "chunks"])):
+    """Container for a hash value and the chunks used when computing it.
+
+    Parameters
+    ----------
+    value : :class:`np.int64`
+        The hash value.
+    chunks : tuple
+        The chunks the array had when the hash was computed.
+    """
+
+    __slots__ = ()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            raise TypeError(f"Unable to compare {repr(self)} to {repr(other)}")
+        if self.chunks != other.chunks:
+            raise ValueError(
+                "Unable to compare arrays with different chunks: "
+                f"{self.chunks} != {other.chunks}"
+            )
+        return self.value == other.value
+
+
+def array_id(array: np.ndarray | da.Array) -> str:
+    """Get a deterministic token representing `array`."""
+    if isinstance(array, np.ma.MaskedArray):
+        # Tokenizing a masked array is much slower than separately tokenizing
+        # the data and mask.
+        return tokenize((tokenize(array.data), tokenize(array.mask)))
+    return tokenize(array)
+
+
+def _compute_hashes(arrays: Iterable[np.ndarray | da.Array]) -> dict[str, _ArrayHash]:
+    """Compute hashes for the arrays that will be compared."""
+
+    def is_numerical(dtype):
+        return np.issubdtype(dtype, np.bool_) or np.issubdtype(dtype, np.number)
+
+    def group_key(a):
+        if is_numerical(a.dtype):
+            dtype = "numerical"
+        else:
+            dtype = str(a.dtype)
+        return a.shape, dtype
+
+    hashes = {}
+
+    arrays = sorted(arrays, key=group_key)
+    for _, group_iter in itertools.groupby(arrays, key=group_key):
+        group = list(group_iter)
+        # Unify dtype for numerical arrays, as the hash depends on it
+        if is_numerical(group[0].dtype):
+            dtype = np.result_type(*group)
+            same_dtype_arrays = [a.astype(dtype) for a in group]
+        else:
+            same_dtype_arrays = group
+        if any(isinstance(a, da.Array) for a in same_dtype_arrays):
+            # Unify chunks as the hash depends on the chunks.
+            indices = tuple(range(group[0].ndim))[::-1]
+            argpairs = [(a, indices) for a in same_dtype_arrays]
+            rechunked_arrays = da.core.unify_chunks(*itertools.chain(*argpairs))[1]
+        else:
+            rechunked_arrays = same_dtype_arrays
+        for array, rechunked in zip(group, rechunked_arrays):
+            hashes[array_id(array)] = (
+                _hash_array(rechunked),
+                rechunked.chunks if isinstance(rechunked, da.Array) else tuple(),
+            )
```

Comment:
Would a correct summary be that this block determines:

- What is part of equality:
  - Shape (obviously)
  - Numerical or not
- What is NOT part of equality:
  - Different numerical dtypes
  - Chunking

A comment highlighting this would be useful.